src/trees/prufer.jl

```diff
@@ -0,0 +1,96 @@
+"""
+    is_tree(g)
+
+Returns true if g is a tree: that is, a simple, connected undirected graph, with nv-1 edges (nv = number of vertices). Trees are the minimal connected graphs; equivalently they have no cycles.  
+
+This function does not apply to directed graphs. Directed trees are sometimes called [polytrees](https://en.wikipedia.org/wiki/Polytree)). 
+
+"""
+
+function is_tree end
+
+@traitfn function is_tree(g::::(!IsDirected))
+    return ne(g) == nv(g) - 1 && is_connected(g)
+end
+
+function _is_prufer(c)
+    return ndims(c) == 1 &&
+           eltype(c) <: Integer &&
+           (isempty(c) || maximum(c) <= length(c) + 2)
+end
+
+function _degree_from_prufer(c::Vector{T})::Vector{T} where {T<:Integer}
+    """
+    Degree sequence from Prüfer code.
+    Returns d such that d[i] = 1 + number of occurences of i in c
+    """
+    n = length(c) + 2
+    return [T(count(==(i), c) + 1) for i in 1:n]
+end
+
+"""
+    prufer_decode(code)
+
+Returns the unique tree associated with the given (Prüfer) code. 
+Each tree of size n is associated with a Prüfer sequence (a[1], ..., a[n-2]) with 1 ⩽ a[i] ⩽ n. The sequence is constructed recursively by the leaf removal algoritm. At step k, the leaf with smallest index is removed and its unique neighbor is added to the Prüfer sequence, giving a[k]. The decoding algorithm goes backward.  The tree associated with the empty sequence is the 2-vertices tree with one edge.
+Ref: [Prüfer sequence on Wikipedia](https://en.wikipedia.org/wiki/Pr%C3%BCfer_sequence)
+
+"""
+
+function prufer_decode(code::AbstractVector{T})::SimpleGraph{T} where {T<:Integer}
+    !_is_prufer(code) && throw(
+        ArgumentError("The code must have one dimension and must be a Prüfer sequence. "),
+    )
+    isempty(code) && return path_graph(T(2)) # the empty Prüfer sequence codes for the one-edge tree
+
+    n = length(code) + 2
+    d = _degree_from_prufer(code)
+    L = BinaryMinHeap{T}(findall(==(1), d))
```

Comment:
```suggestion
    L = BinaryMaxHeap{T}(findall(==(1), d))
```
This should be a max-heap, right?

Otherwise I think this will not work for the path graph `2 -- 4 -- 3 -- 1` as this graph has the Prüfer sequence `[4, 3]` and that will then result in the path graph `1 -- 4 -- 3 -- 2`.

Reply:
Prüfer encoding relies on the choice of a leaf selection rule at each step. Some papers use the minimum leaf rule, others use the maximum leaf rule. Different selection rules with result in different Prüfer encodings.

In the docstrings I chose the minimum leaf rule (I think it's the conventional choice), so at each step I remove the leaf with smallest index, hence my use of `BinaryMinHeap` (the `pop!` method removes the root, ie the smallest element).

Reply:
That being said, maybe we could add the selection rule as a parameter.

Reply:
A right, my fault, I calculated the Prüfer sequence incorrect, when trying to understand the algorithm. Then we probably won't need the import of `BinaryMaxHeap`?

Reply:
Haha, at the beginning I wasn't sure if I was going to use the min-rule or max-rule, so I imported both... what do you think about having an argument `selection_rule` that could be `"min"` (default) or `"max"` ?

Reply:
I am not sure how much such a functionality is used - but what you could also do is to have an optional  `ordering` parameter that takes a `Base.Order.Ordering` object as described [here](https://docs.julialang.org/en/v1/stdlib/Test/#Test.@inferred). 

As far as I know, the `BinaryHeap` class from `DataStructures.jl` should work with that.

Reply:
Ok. I really think that's unnecessary for the moment and I'll leave this for later.